src/components/helpers/Pagination.tsx

```diff
@@ -0,0 +1,175 @@
+import { FormEvent, useEffect, useState } from 'react';
+import {
+  HiChevronDoubleLeft,
+  HiChevronDoubleRight,
+  HiChevronLeft,
+  HiChevronRight,
+  HiDotsHorizontal,
+} from 'react-icons/hi';
+import { MdKeyboardArrowRight } from 'react-icons/md';
+import useRouter from './useRouter';
+
+interface PaginationProps {
+  totalItems: number;
+}
+
+function Pagination({ totalItems }: PaginationProps) {
+  const [currentPage, setCurrentPage] = useState(1);
+  const [totalPages, setTotalPages] = useState(0);
+  const [itemsPerPage, setItemsPerPage] = useState(12);
+  const [futureCurrentPage, setFutureCurrentPage] = useState<number | null>(null);
+  const { pathname, push } = useRouter();
+
+  useEffect(() => {
+    const pages = Math.ceil(totalItems / itemsPerPage);
+    setTotalPages(pages);
+    push(`${pathname}?page=${currentPage}&itemsPerPage=${itemsPerPage}`);
+  }, [currentPage, totalPages, itemsPerPage, totalItems, push, pathname]);
+
+  const goToFirstPage = () => setCurrentPage(1);
+  const goToPrevPage = () => setCurrentPage(currentPage - 1);
+  const goToNextPage = () => setCurrentPage(currentPage + 1);
+  const goToLastPage = () => setCurrentPage(totalPages);
+
+  const last5Pages = [totalPages - 4, totalPages - 3, totalPages - 2, totalPages - 1, totalPages];
+
+  const goToPage = (e: FormEvent) => {
+    e.preventDefault();
+    if (futureCurrentPage) setCurrentPage(futureCurrentPage);
+  };
+
+  const singlePageNumber = (page: number) => {
+    return (
+      <button
+        className={`pagination-btn number ${currentPage === page ? 'active' : null}`}
+        key={page}
+        type="button"
+        onClick={() => setCurrentPage(page)}
+        onKeyDown={() => setCurrentPage(page)}
+      >
+        {page}
+      </button>
+    );
+  };
+
+  const pages = () => {
+    if (totalPages < 7) {
+      return (
+        <>
+          {Array(totalPages)
+            .fill(0)
+            .map((_, page) => singlePageNumber(page + 1))}
+        </>
+      );
+    }
+    if (totalPages >= 7 && currentPage < 3) {
+      return (
+        <>
+          {[1, 2, 3, 4, 5].map((page) => singlePageNumber(page))}
+          <HiDotsHorizontal className="dots" />
+          {singlePageNumber(totalPages)}
+        </>
+      );
+    }
+    if (totalPages >= 7 && currentPage >= 3 && currentPage <= totalPages - 2) {
+      return (
+        <>
+          {singlePageNumber(1)}
+          <HiDotsHorizontal className="dots" />
+          {singlePageNumber(currentPage - 1)}
+          {singlePageNumber(currentPage)}
+          {singlePageNumber(currentPage + 1)}
+          <HiDotsHorizontal className="dots" />
+          {singlePageNumber(totalPages)}
+        </>
+      );
+    }
+    if (totalPages >= 7 && currentPage >= totalPages - 2) {
+      return (
+        <>
+          {singlePageNumber(1)}
+          <HiDotsHorizontal className="dots" />
+          {last5Pages.map((page) => singlePageNumber(page))}
```

Comment:
Hm... totalPages.slice(-5) i mapować?

Reply:
Hm, zastanawiam się, czy tego nie byłoby lepiej tego zmapować. Nawet zwykłym forem, zacząć od pages.length i iść o 1 w dół. Bo teraz jak masz mniej stron, to w arajce masz liczby ujemne, których i tak musisz się pozbyć / nie wyświetlać ich, a jak pojedziesz metodą / pętlą / czymś odrobinę bardziej zaawansowanym, to możesz od razu się takich cyfr pozbyć.